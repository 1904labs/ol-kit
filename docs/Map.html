<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>ol-kit | Map</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0, user-scalable=no, target-densityDpi=device-dpi" />


    <!-- these scripts must be at the top to properly syntax highlight code examples -->
    <script src="https://cdn.jsdelivr.net/gh/google/code-prettify@master/loader/run_prettify.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/babel-standalone/6.26.0/babel.min.js" integrity="sha256-FiZMk1zgTeujzf/+vomWZGZ9r00+xnGvOgXoj0Jo1jA=" crossorigin="anonymous"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.6.3/css/all.css" integrity="sha384-UHRtZLI+pbxtHCWp1t77Bi1L4ZtiqrqD80Kn4Z8NTSRyMA2Fd33n5dQ8lWUE00s/" crossorigin="anonymous">
    <link type="text/css" rel="stylesheet" href="https://jmblog.github.io/color-themes-for-google-code-prettify/themes/tomorrow.min.css">
    <link type="text/css" rel="stylesheet" href="styles/app.css">
    <link type="text/css" rel="stylesheet" href="styles/iframe.css">
</head>

<body>
    <div id="stickyNavbarOverlay"></div>
    <div class="top-navbar">
        <div class="container">
            <nav class="navbar" role="navigation" aria-label="main navigation">
                <div class="navbar-brand">
                    <a href="index.html"><img src="static/ol-kit-logo.png" alt="logo"></a>
                    <a id="hamburger" role="button" class="navbar-burger" aria-label="menu" aria-expanded="false">
                        <span aria-hidden="true"></span>
                        <span aria-hidden="true"></span>
                        <span aria-hidden="true"></span>
                    </a>
                </div>
                 
                <div class="navbar-menu">
                    <div class="navbar-end">
                    
                        <div class="navbar-item">
                            <a href="tutorial-Getting%20Started.html">
                                Get Started
                            </a>
                        </div>
                    
                        <div class="navbar-item">
                            <a href="docs.html">
                                Docs
                            </a>
                        </div>
                    
                        <div class="navbar-item">
                            <a href="https://github.com/MonsantoCo/ol-kit">
                                Github
                            </a>
                        </div>
                    
                    </div>
                </div>
                
            </nav>
        </div>
    </div>
    <div id="page-wrapper" class="container">
        <div class="columns">
            <div class="column is-3" id="sidebar-nav">
                <div class="sidebar sidebar__inner">
                    <nav>
                        <div class="category"><h3>Tutorials</h3><ul><li><a class="undefined"href="tutorial-Getting Started.html">Getting Started</a></li><li><a class="undefined"href="tutorial-Popup_.html">Popup</a></li><li><a class="undefined"href="tutorial-connectToMap.html">connectToMap</a></li><li><a class="undefined"href="tutorial-i18n Support.html">i18n Support</a></li></ul></div><div class="category"><h2>Basemap</h2><h3>Components</h3><ul><li><a href="BasemapManager.html">BasemapManager</a></li><li><a href="BingMaps.html">BingMaps</a></li><li><a href="BlankWhite.html">BlankWhite</a></li><li><a href="OpenStreetMap.html">OpenStreetMap</a></li><li><a href="StamenTerrain.html">StamenTerrain</a></li><li><a href="StamenTonerDark.html">StamenTonerDark</a></li><li><a href="StamenTonerLite.html">StamenTonerLite</a></li></ul></div><div class="category"><h2>Classes</h2><h3>Classes</h3><ul><li><a href="VectorLayer.html">VectorLayer</a></li></ul></div><div class="category"><h2>Controls</h2><h3>Components</h3><ul><li><a href="Compass.html">Compass</a></li><li><a href="Controls.html">Controls</a></li><li><a href="ZoomControls.html">ZoomControls</a></li></ul></div><div class="category"><h2>Map</h2><h3>Components</h3><ul><li><a href="Map.html">Map</a></li></ul><h3>Utils</h3><ul><li><a href="global.html#centerAndZoom">centerAndZoom</a></li><li><a href="global.html#connectToMap">connectToMap</a></li><li><a href="global.html#createMap">createMap</a></li><li><a href="global.html#createSelectInteraction">createSelectInteraction</a></li><li><a href="global.html#updateMapFromUrl">updateMapFromUrl</a></li><li><a href="global.html#updateUrlFromMap">updateUrlFromMap</a></li></ul></div><div class="category"><h2>Popup</h2><h3>Components</h3><ul><li><a href="Popup.html">Popup</a></li><li><a href="PopupActionGroup.html">PopupActionGroup</a></li><li><a href="PopupActionItem.html">PopupActionItem</a></li><li><a href="PopupBase.html">PopupBase</a></li><li><a href="PopupDataList.html">PopupDataList</a></li><li><a href="PopupDefaultInsert.html">PopupDefaultInsert</a></li><li><a href="PopupDefaultPage.html">PopupDefaultPage</a></li><li><a href="PopupPageLayout.html">PopupPageLayout</a></li><li><a href="PopupPageLayoutChild.html">PopupPageLayoutChild</a></li><li><a href="PopupTabs.html">PopupTabs</a></li></ul><h3>Utils</h3><ul><li><a href="global.html#addMovementListener">addMovementListener</a></li><li><a href="global.html#calculateViewPadding">calculateViewPadding</a></li><li><a href="global.html#getLayersAndFeaturesForEvent">getLayersAndFeaturesForEvent</a></li><li><a href="global.html#getPopupPositionFromFeatures">getPopupPositionFromFeatures</a></li><li><a href="global.html#removeMovementListener">removeMovementListener</a></li></ul></div>
                    </nav>
                </div>
            </div>
            <div class="column is-9-desktop" id="main-content-wrapper">
                <div class="content">
                    




<section>

<header>
    
        <h2>
            &ltMap /&gt
        </h2>
        
            <div class="class-description">A Reactified ol.Map wrapper component</div>
        
    
</header>

<article>
    <div class="container-overview">
    
        
            <div class='vertical-section'>
                <h3 class="title"><span>Constructor</span></h3>
                <div class="members">
                    <div class="member">


    
    <h4 class="name" id="Map">
        <a class="href-link" href="#Map">#</a>
        
        <span class="code-name">
            
                &ltMap <i>children</i> <i>fullScreen</i> <i>logoPosition</i> <i>map</i> <i>onMapInit</i> <i>updateUrlDebounce</i> <i>updateUrlFromView</i> <i>updateViewFromUrl</i> <i>urlViewParam</i> <i>selectInteraction</i> <i>style</i> <i>translations</i> /&gt
            
        </span>
    </h4>
    

    
    







    
        <h5>PropTypes:</h5>
        

<table class="params">
  <thead>
  <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Required</th>
      <th>Description</th>
      <th class="last">Default</th>
  </tr>
  </thead>

  <tbody>
  
        <tr class="deep-level-0">
          <td class="name"><code>children</code></td>
          <td>
            union
          </td>
          <td>No</td>
          <td>any ol-kit children components will automatically be passed a reference to the map object via the `map` prop</td>
          <td></td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>fullScreen</code></td>
          <td>
            bool
          </td>
          <td>No</td>
          <td>if this is set to false, the map will fill it's parent container</td>
          <td>false</td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>logoPosition</code></td>
          <td>
            string
          </td>
          <td>No</td>
          <td>place the ol-kit logo on the 'left', 'right', or set to 'none' to hide</td>
          <td>'left'</td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>map</code></td>
          <td>
            object
          </td>
          <td>No</td>
          <td>optionally pass a custom map</td>
          <td>null</td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>onMapInit</code></td>
          <td>
            func
          </td>
          <td>No</td>
          <td>callback called with initialized map object after optional animations complete
    note: if a Promise is returned from this function, Map will wait for onMapInit to resolve before rendering children</td>
          <td>() => {}</td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>updateUrlDebounce</code></td>
          <td>
            number
          </td>
          <td>No</td>
          <td>the length of debounce on map movements before the url gets updated</td>
          <td>400</td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>updateUrlFromView</code></td>
          <td>
            bool
          </td>
          <td>No</td>
          <td>add map location coords + zoom level to url as query params</td>
          <td>true</td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>updateViewFromUrl</code></td>
          <td>
            bool
          </td>
          <td>No</td>
          <td>update map view based off the url param</td>
          <td>true</td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>urlViewParam</code></td>
          <td>
            string
          </td>
          <td>No</td>
          <td>change the url param used to set the map location coords</td>
          <td>'view'</td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>selectInteraction</code></td>
          <td>
            object
          </td>
          <td>No</td>
          <td>an openlayers select interaction passed down to connected components - created internally if not passed as prop</td>
          <td></td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>style</code></td>
          <td>
            object
          </td>
          <td>No</td>
          <td>apply inline styles to the map container</td>
          <td>{}</td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>translations</code></td>
          <td>
            object
          </td>
          <td>No</td>
          <td>object of string key/values (see: locales)</td>
          <td>function()</td>
        </tr>
        
  
  </tbody>
</table>
    







<dl class="details">

    

    
    <dt class="tag-since">Since:</dt>
    <dd class="tag-since"><ul class="dummy"><li>0.1.0</li></ul></dd>
    

    

    

    

    

    

    

    

    

    

    
    

    

    

    
        <p class="tag-source">
            <a href="Map_Map.js.html" class="button">View Source</a>
            <span>
                <a href="Map_Map.js.html">Map/Map.js</a>, <a href="Map_Map.js.html#line22">line 22</a>
            </span>
        </p>
    
</dl>





















</div>
                </div>
            </div>
        
    
    </div>
    
        
    
    

    

    

    

    

    

    

    

    

    
</article>

</section>




                </div>
            </div>
        </div>
    </div>
<footer class="footer">
    <div class="content has-text-centered">
        <div style="margin: 0 auto; text-align: center; background: white; margin: 30px 0;">
            <a href="https://www.cropscience.bayer.com" target="_blank">
                <img src="static/bayer-logo.png" width="200px" />
            </a>
            <div>&copy; 2020 Bayer Crop Science; All Rights Reserved.</div>
        </div>
    </div>
</footer>

<script src="https://cdnjs.cloudflare.com/ajax/libs/zepto/1.2.0/zepto.min.js" integrity="sha256-vrn14y7WH7zgEElyQqm2uCGSQrX/xjYDjniRUQx3NyU=" crossorigin="anonymous"></script>
<script src="https://unpkg.com/sticky-sidebar/dist/sticky-sidebar.min.js"></script>
<script src="scripts/app.js"></script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>