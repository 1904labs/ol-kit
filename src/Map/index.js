export { default as Map } from './Map'
export * from './utils'
