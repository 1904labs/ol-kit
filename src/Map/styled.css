.logoContainer { 
  width: 100%;
  position: absolute;
  display: flex;
  align-items: flex-end;
  bottom: 1px;
  z-index: 1;
}

/* justify-content: ${props => props.position === 'right' ? 'flex-end' : 'flex-start'}; */

.logo {
  width: 15px;
  height: 15px;
  padding: 2px;
  cursor: pointer;
}

.logoText {
  font-size: 11px;
  font-family: Roboto,Arial,sans-serif;
  padding-left: 2px;
  margin: 0px;
}

.styledMap {
  height: 100%;
  width: 100%;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
}

/* position: ${p => p.fullScreen ? 'fixed' : 'relative'}; */
