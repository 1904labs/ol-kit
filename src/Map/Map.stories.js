import React from 'react'
import Map from './Map'

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction
export default {
  title: 'Example/Map',
  component: Map,
  tags: ['autodocs'],
  argTypes: {
    isMiltiMap: {
      control: {
        type: 'boolean',
        default: false
      }
    },
    logoPosition: {
      control: {
        type: 'select',
        options: ['left', 'right', 'none'],
        default: 'none'
      }
    },
    fullscreen: {
      control: {
        type: 'boolean',
        default: false
      }
    }
  },
  decorators: [
    (Story) => (
      <div style={{ width: '100%', height: '500px' }}>
        <Story />
      </div>
    )
  ]
}

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Primary = {
  args: {
    isMiltiMap: false,
    logoPosition: 'none',
    fullscreen: false
  }
}

// export const Secondary = {
//   args: {
//     label: 'Button',
//   },
// };

// export const Large = {
//   args: {
//     size: 'large',
//     label: 'Button',
//   },
// };

// export const Small = {
//   args: {
//     size: 'small',
//     label: 'Button',
//   },
// };
