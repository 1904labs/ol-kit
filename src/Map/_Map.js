import React from 'react'
import Map from './Map' // this should be the only relative ./Map import in ol-kit
import { Provider } from 'Provider'
import { MultiMapContext } from 'MultiMapManager'
import { ErrorBoundary } from 'ErrorBoundary'

// check for MultiMapManager so Provider wrapper does not also render a context
function _Map (props) {
  return (
    !MultiMapContext
      ? (
        <ErrorBoundary floating={true}>
          <Provider {...props}>
            <Map {...props} />
          </Provider>
        </ErrorBoundary>
      )
      : <Map {...props} />
  )
}

export default _Map
