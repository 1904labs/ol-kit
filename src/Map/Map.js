import React from 'react'
import PropTypes from 'prop-types'
import nanoid from 'nanoid'
import debounce from 'lodash.debounce'
import { StyledMap } from './styled'
import { createMap, updateMapFromUrl, updateUrlFromMap } from './utils'
import ugh from 'ugh'
import en from 'locales/en'

// context should only be created when <Map> is mounted (see constructor), otherwise it's null so child comps don't use context
export let MapContext = null

/**
 * A Reactified ol.Map wrapper component
 * @component
 * @category Map
 * @since 0.1.0
 */
class Map extends React.Component {
  constructor (props) {
    super(props)

    // this is used to create a unique identifier for the map div
    this.target = `_ol_kit_map_${nanoid(6)}`

    // create a map context
    MapContext = React.createContext()
  }

  componentDidMount () {
    const { map, onMapInit, updateUrlDebounce, updateUrlFromView, updateViewFromUrl, urlViewParam } = this.props

    if (!map) {
      // if no map was passed, create the map
      this.map = createMap({ target: this.target })
      // force update to get this.map into the context
      this.forceUpdate()
    } else {
      this.map = map
    }

    if (updateUrlFromView) {
      const setUrl = () => updateUrlFromMap(this.map, urlViewParam)
      const mapMoveListener = debounce(setUrl, updateUrlDebounce)

      // update the url param after map movements
      this.map.on('moveend', mapMoveListener)
    }

    if (updateViewFromUrl) {
      // read the url to update the map from view info
      updateMapFromUrl(this.map, urlViewParam)
        .catch(ugh.error)
        .finally(() => onMapInit(this.map)) // always fire callback with map reference on success/failure
    } else {
      // callback that returns a reference to the created map
      onMapInit(this.map)
    }
  }

  getContextValue = () => {
    return {
      map: this.map,
      translations: en
    }
  }

  render () {
    const { children, fullScreen } = this.props

    return (
      <>
        <StyledMap
          id={this.target}
          fullScreen={fullScreen} />
        <MapContext.Provider value={this.getContextValue()}>
          {!this.map // wait for a map to exist before rendering children that need a ref to map
            ? null
            : children}
        </MapContext.Provider>
      </>
    )
  }
}

Map.defaultProps = {
  fullScreen: false,
  map: null,
  onMapInit: () => {},
  updateUrlDebounce: 400,
  updateUrlFromView: true,
  updateViewFromUrl: true,
  urlViewParam: 'view'
}

Map.propTypes = {
  /** any ol-kit children components will automatically be passed a reference to the map object via the `map` prop */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  /** if this is set to false, the map will fill it's parent container */
  fullScreen: PropTypes.bool,
  /** optionally pass a custom map */
  map: PropTypes.object,
  /** returns an initialized map object after optional animations complete */
  onMapInit: PropTypes.func,
  /** the length of debounce on map movements before the url gets updated */
  updateUrlDebounce: PropTypes.number,
  /** add map location coords + zoom level to url as query params */
  updateUrlFromView: PropTypes.bool,
  /** update map view based off the url param */
  updateViewFromUrl: PropTypes.bool,
  /** change the url param used to set the map location coords */
  urlViewParam: PropTypes.string
}

export default Map
