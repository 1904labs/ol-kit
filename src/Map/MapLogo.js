import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { Logo } from './styled'
import OL_KIT_MARK from 'images/ol_kit_mark.svg'
import OL_KIT_MARK_BLACK from 'images/ol_kit_mark_black.svg'

export default function MapLogo (props) {
  const { mapLogoPosition } = props
  const [isHovered, setHovered] = useState(false)

  return (
    mapLogoPosition === 'none'
      ? null
      : (
        <Logo
          position={mapLogoPosition}
          onMouseEnter={() => setHovered(true)}
          onMouseLeave={() => setHovered(false)}
          title='Powered by ol-kit'
          href='https://ol-kit.com/'
          target='_blank'>
          {isHovered
            ? <OL_KIT_MARK />
            : <OL_KIT_MARK_BLACK />
          }
        </Logo>
      )
  )
}

MapLogo.defaultProps = {
  mapLogoPosition: 'left'
}

MapLogo.propTypes = {
  mapLogoPosition: PropTypes.string
}
