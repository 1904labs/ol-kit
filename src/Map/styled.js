import styled from 'styled-components'

export const Logo = styled.a.attrs(props => ({
  title: props.title
}))`
  position: absolute;
  width: 28px;
  bottom: 8px;
  z-index: 1;
  cursor: pointer;
  ${p => p.position === 'left' ? 'left: 14px;' : 'right: 14px;'}
`

export const StyledMap = styled.div`
  height: 100%;
  width: 100%;
  position: ${p => p.fullScreen ? 'fixed' : 'relative'};
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  ${p => p.style};
`
