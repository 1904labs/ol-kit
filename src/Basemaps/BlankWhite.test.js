import React from 'react'
import { shallow, mount } from 'enzyme'
import { cleanup, fireEvent, getByText, render, waitFor } from '@testing-library/react'
import { prettyDOM } from '@testing-library/dom'
import Map from '../Map'
import BlankWhite from './BlankWhite'
import olMap from 'ol/map'
import olLayerVector from 'ol/layer/vector'

describe('<BlankWhite />', () => {
  afterEach(cleanup)
  it('should render a basic basemap option component', async () => {
    const { container } = render(<Map><BlankWhite /></Map>)

    // wait for async child render
    await waitFor(() => {}, { container })

    expect(prettyDOM(container)).toMatchSnapshot()
  })
  it('should add a basemap to an empty map when clicked', async () => {
    const map = new olMap()
    const { container } = render(<Map map={map}><BlankWhite /></Map>)

    // wait for async child render
    await waitFor(() => {}, { container })

    expect(map.getLayers().getArray().length).toBe(0)
    fireEvent.click(getByText(container, 'Blank White'))
    expect(map.getLayers().getArray().length).toBe(1)
  })

  it('should set the first layer to a basemap to a map containing a preexisting basemap when clicked with a string layerTypeID.', async () => {
    const mockLayerTypeID = 'mockLayerTypeID'
    const mockLayer = new olLayerVector()

    mockLayer.set(mockLayerTypeID, 'osm')

    const map = new olMap({
      layers: [
        mockLayer
      ]
    })
    const { container } = render(<Map map={map}><BlankWhite layerTypeID={mockLayerTypeID} /></Map>)

    // wait for async child render
    await waitFor(() => {}, { container })

    expect(map.getLayers().getArray().length).toBe(1)
    fireEvent.click(getByText(container, 'Blank White'))
    expect(map.getLayers().getArray().length).toBe(1)
  })

  it('should fire the callback when the layers are changed', async () => {
    const callback = jest.fn()
    const { container } = render(<Map><BlankWhite onBasemapChanged={callback} /></Map>)

    // wait for async child render
    await waitFor(() => {}, { container })

    expect(callback).not.toHaveBeenCalled()
    fireEvent.click(getByText(container, 'Blank White'))
    expect(callback).toHaveBeenCalledTimes(1)
  })

  it('should render a blue border to indicate when the layer is present on the map', async () => {
    const callback = jest.fn()
    const onMapInit = jest.fn()
    const wrapper = mount(<Map onMapInit={onMapInit}><BlankWhite onBasemapChanged={callback} /></Map>)

    // wait for async child render
    await waitFor(() => expect(onMapInit).toHaveBeenCalled())
    wrapper.update()

    expect(wrapper.find('._ol_kit_basemapOption').first().prop('isActive')).toBeFalsy()
    wrapper.find('._ol_kit_basemapOption').first().simulate('click')
    expect(callback).toHaveBeenCalledTimes(1)
    expect(wrapper.find('._ol_kit_basemapOption').first().prop('isActive')).toBeTruthy()
  })
})
