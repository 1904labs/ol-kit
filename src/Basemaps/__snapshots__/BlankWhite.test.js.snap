// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BlankWhite /> should render a basic basemap option component 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
